### PR TITLE
[#11267] Logs interface: Support more filters

### PR DESCRIPTION
**Pull request by ypinhsuan**
Part of #11267

**Outline of Solution**
This PR contains the frontend part.
Users can add more filters under the `Advanced Filters` section. Filters will be added automatically when users click on the field. Updated the search form to support querying by logs event type as well. 
m2

The API endpoint is a dropdown list containing action class names. Users have to type for all other fields. Trace ID, Source location file, Source location function, User Google ID, User Registration Key and User Email can be filled by clicking the corresponding fields in the logs below.
image

Histogram view for error logs will be in a separate PR.

### CODE REVIEW

File: src/web/app/pages-logs/logs-page.component.html
Comment (wkurniawan07) [inline src/web/app/pages-logs/logs-page.component.html:97]: HTML should not have duplicate IDs

File: src/web/app/pages-logs/logs-page.component.ts
Comment (wkurniawan07) [inline src/web/app/pages-logs/logs-page.component.ts:130]: I don't think you should combine severity level and event. Even combining severity and min severity is debatable.

File: src/web/app/components/logs-table/logs-table.component.html
Comment (wkurniawan07) [inline src/web/app/components/logs-table/logs-table.component.html:40]: Based on this and also the screenshot the filter is added upon clicking. Not right/wrong on its own, but it's not the most intuitive manner, especially since nothing in this displayed value indicates that there is an interaction when clicking (text doesn't look different, cursor doesn't change into pointer, etc; the highlight only appears on hover and it doesn't tell anything about being clickable).

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/ui/constants/ActionNames.java
Comment (ypinhsuan) [reply src/main/java/teammates/ui/constants/ActionNames.java:12]: I will define the strings here and remove those in `Const`

File: src/web/app/pages-logs/logs-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-logs/logs-page.component.ts:130]: Will separate them into 3 different error messages

File: src/web/app/components/logs-table/logs-table.component.html
Comment (ypinhsuan) [reply src/web/app/components/logs-table/logs-table.component.html:40]: I have added a tooltip to let users know that the value can be clicked. The cursor will also change to pointer when hover over fields that can be clicked.
image

File: src/main/java/teammates/ui/constants/ActionNames.java
Comment (wkurniawan07) [inline src/main/java/teammates/ui/constants/ActionNames.java:8]: Even though this works, this adds an unnecessary coupling to the list of actions already defined somewhere else. If we decide to add new action, rename some actions, deprecate/remove some actions, this enum's accuracy is compromised if not kept in sync. 

Better alternative: create an admin-only (later on, admin+maintainer-only) API endpoint that returns the list of actions. It is straightforward to get this list populated from the other place that already listed this down. (hint: you're looking at the right place if you find out that you need to (1) change a field from `private` to package level and (2) flatten a nested map)

File: src/web/app/pages-logs/logs-page.component.ts
Comment (wkurniawan07) [inline src/web/app/pages-logs/logs-page.component.ts:87]: I know this is not caused by this PR, but is there a reason for naming this field `previous`? Is there a `currentQueryParams` or `nextQueryParams` somewhere, or planned to be?

File: src/web/services/log.service.ts
Comment (wkurniawan07) [inline src/web/services/log.service.ts:100]: I believe this has been renamed to `actionclass`

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/web/app/pages-logs/logs-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-logs/logs-page.component.ts:87]: When users load more logs, the request parameters need to be the same as the previous request. I named it this way because this field stores the previous request and use it for the current request. But since there is no `currentQueryParams` or `nextQueryParams`, I will change it to just `queryParams`

File: src/web/services/log.service.ts
Comment (ypinhsuan) [reply src/web/services/log.service.ts:100]: Will update this

File: src/main/java/teammates/ui/constants/ActionNames.java
Comment (ypinhsuan) [reply src/main/java/teammates/ui/constants/ActionNames.java]: I have removed this file and action names are now retrieved by flattening the map in ActionFactory. The backend code will be added to #11269

File: src/web/services/log.service.ts
Comment (wkurniawan07) [inline src/web/services/log.service.ts:62]: This can be `QueryParams` type?

File: src/web/app/pages-logs/logs-page.component.ts
Comment (wkurniawan07) [inline src/web/app/pages-logs/logs-page.component.ts:55]: Looks like a lot of similarity between this and `SearchLogsFormModel`

File: src/web/app/pages-logs/logs-page.component.ts
Comment (wkurniawan07) [inline src/web/app/pages-logs/logs-page.component.ts:41]: I'm referring to the header comment not displayed in the diff:
```
Query parameters for HTTP request
```
This is way too generic. The query params here is obviously only for one specific endpoint, namely the logs endpoint.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Almost there

File: src/web/services/log.service.ts
Comment (ypinhsuan) [reply src/web/services/log.service.ts:62]: Will update this

File: src/web/app/pages-logs/logs-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-logs/logs-page.component.ts:55]: I will remove the duplicate fields in `formModel`. The fields will be added to `queryParams` directly instead of going through the `formModel`.

File: src/web/app/pages-logs/logs-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-logs/logs-page.component.ts:41]: ok! will specify that it is for logs endpoint.

File: src/web/app/pages-logs/logs-page.component.ts
Comment (wkurniawan07) [inline src/web/app/pages-logs/logs-page.component.ts:33]: If this interface is `export`ed, it becomes important to name it to a less generic name. Otherwise something may end up importing this interface, thinking it is a generic query parameter interface, only to find out otherwise.

File: src/web/app/pages-logs/logs-page.component.ts
Comment (wkurniawan07) [inline src/web/app/pages-logs/logs-page.component.ts:177]: This method need to be renamed or better explained. Just looking at method name `setQueryParam` and two timestamp parameters tells nothing about some other fields being changed to `undefined` or other values.

File: src/web/services/log.service.ts
Comment (wkurniawan07) [inline src/web/services/log.service.ts:3]: A service class should not be aware of component class. But the front-end architecture right now is sketchy at best, so this can stay as is.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/web/app/pages-logs/logs-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-logs/logs-page.component.ts:33]: I will change the name of interface to a more specific one.

File: src/web/services/log.service.ts
Comment (ypinhsuan) [reply src/web/services/log.service.ts:3]: Would it be better if I put this interface in the service class and the component imports the interface?

File: src/web/app/pages-logs/logs-page.component.ts
Comment (ypinhsuan) [reply src/web/app/pages-logs/logs-page.component.ts:47]: Binding data directly to `formModel` leads to some problem. When user changes the filters but do not press the query button, the `queryParams` is still modified. This results in error when user loads more logs as same params are expected. So I will create another interface `advancedFilters` to reduce the duplication instead.

File: src/web/services/log.service.ts
Comment (wkurniawan07) [reply src/web/services/log.service.ts:3]: Sounds like a better alternative to me

File: src/web/app/pages-logs/logs-page.component.spec.ts
Comment (wkurniawan07) [inline src/web/app/pages-logs/logs-page.component.spec.ts:174]: Typo in `message`; happens a few times in this file

**Review by wkurniawan07 — COMMENTED** [review]
Good to go after one minor adjustment

File: src/web/app/pages-logs/logs-page.component.spec.ts
Comment (ypinhsuan) [reply src/web/app/pages-logs/logs-page.component.spec.ts]: Fixed!

**Review by wkurniawan07 — APPROVED** [review]
LGTM

**Comment by wkurniawan07** [comment]
@ypinhsuan small request: can you add the link to the logs page from the admin navbar?

**Comment by ypinhsuan** [comment]
I have added the link in nav bar

**Pull request merged**
